js/start.js

```diff
@@ -47,59 +34,192 @@ app.loadingModal = new LoadingModal({
   removeOnRoute: false,
 }).render().open();
 
-// get the server config
-$.get(app.getServerUrl('ob/config')).done((data) => {
-  app.profile = new Profile({ id: data.guid });
+const fetchConfigDeferred = $.Deferred();
 
-  // todo: for now busting cache on fetch pending
-  // issue where my server is not running the latest
-  // code which solves this.
-  app.profile.fetch({ cache: false })
-    .done(() => {
-      app.pageNav.navigable = true;
-      Backbone.history.start();
-      app.loadingModal.close();
-    }).fail((jqXhr) => {
-      if (jqXhr.status === 400) {
-        // for now we'll consider 400 - Bad Request to mean
-        // onboarding is needed. After some pending server changes
-        // 404 should mean onboarding is needed.
-        // todo: follow up with cpacia to ensure the server change is
-        // made.
-
-        // for now we'll just manually save the profile with
-        // some default / dummy values. Later, we'll make the
-        // onboarding modal.
-
-        const profileSave = app.profile.save({}, {
-          type: 'POST',
-        });
-
-        if (!profileSave) {
-          throw new Error('Client side validation failed on your new Profile model.' +
-            'Ensure your defaults are valid.');
-        } else {
-          profileSave.done(() => {
-            app.pageNav.navigable = true;
-            Backbone.history.start();
-            app.loadingModal.close();
-          }).fail((failData) => {
-            app.loadingModal.close();
-            new SimpleMessageModal()
-              .render()
-              .open('Unable To Save Profile', failData.reason || '');
-          });
+function fetchConfig() {
+  $.get(app.getServerUrl('ob/config')).done((...args) => {
+    fetchConfigDeferred.resolve(...args);
+  }).fail(() => {
+    const retryConfigDialog = new Dialog({
+      title: 'Unable to get your server config.',
+      buttons: [{
+        text: 'Retry',
+        fragment: 'retry',
+      }],
+      dismissOnOverlayClick: false,
+      dismissOnEscPress: false,
+      showCloseButton: false,
+    }).on('click-retry', () => {
+      retryConfigDialog.close();
+
+      // slight of hand to ensure the loading modal has a chance to at
+      // least briefly show before another potential failure
+      setTimeout(() => {
+        fetchConfig();
+      }, 300);
+    })
+    .render()
+    .open();
+  });
+
+  return fetchConfigDeferred.promise();
+}
+
+const onboardingNeededDeferred = $.Deferred();
+let profileFetch;
+let settingsFetch;
+let onboardProfile = false;
+let onboardSettings = false;
+let profileFailed;
+let settingsFailed;
+
+function isOnboardingNeeded() {
```

Comment:
If either the settings or profile are not set, we will onboard. While in 99% of cases, the lack of a profile would also indicate the lack of settings, there are some edge cases where one may have been saved but the other failed.

Anyhow, the idea is that those models are critical to the app and the user will not make it out of our start-up flow unless those are set.